Add option --retry

jmgarnier opened the pull request:
Closes #9 

- [x] Retry 3 times to contact Travis when error 500
- [x] Add option `--retry`
- [x] Change behavior to not to stop builds if requets to Travis keep failing
- [ ] Testing it in the wild real world (private project), using https://rubygems.org/gems/jmgarnier_travis_dedup

jmgarnier commented:
Forgot to create a branch locally, I can close this PR if needed and re-opened another one

grosser commented on lib/travis_dedup.rb:
```diff
@@ -5,6 +5,8 @@
 module TravisDedup
   ACTIVE = %w[created started queued]
 
+  DEFAULT_MAX_ATTEMPTS_WHEN_TRAVIS_DOES_NOT_RESPOND = 3
```
`RETRIES = 3` would be enough ;)

grosser commented on lib/travis_dedup.rb:
```diff
@@ -98,17 +100,36 @@ def host
+    end
+
+    def max_attempts
+      DEFAULT_MAX_ATTEMPTS_WHEN_TRAVIS_DOES_NOT_RESPOND
```
not sure why this is a method

jmgarnier replied:
next commit was going to introduce a `--retry` option and change this method. Will amend it.

grosser commented on lib/travis_dedup.rb:
```diff
@@ -98,17 +100,36 @@ def host
+      attempts = 0
+      begin
+        faraday_send(method, path, params, headers)
+      rescue TravisError500 => error
```
doubt this is called TravisError500 oO

grosser replied:
... how about rescueing Faraday::Error instead

grosser commented on lib/travis_dedup.rb:
```diff
@@ -98,17 +100,36 @@ def host
       case response.status
       when 200 then JSON.parse(response.body)
       when 204 then nil
+      when 500
```
I though faraday raises on 500 ... so won't get here ?

jmgarnier replied:
It will get there. Faraday does not raise on 500

grosser commented:
FYI might not want to spend time here since travis is releasing dedup as a feature in Q1

jmgarnier commented:
Thanks for the review.

Following your previous comment on the isue, I asked them about it. Their
feature does a little bit less than travis_dedup so I'd rather move forward
on this PR.

Main difference  with travis_dedup is that it does not delete "Active"
(running) jobs and we really value that behavior.

A screenshot about Travis feature:
[image: Inline image 1]

On Fri, Nov 18, 2016 at 5:10 PM, Michael Grosser notifications@github.com
wrote:

> FYI might not want to spend time here since travis is releasing dedup as a
> feature in Q1
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/grosser/travis_dedup/pull/11#issuecomment-261570374,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAAfcyTpby6-IcYNIMYWcfah3z1dtc01ks5q_c4NgaJpZM4K2bQV
> .

jmgarnier commented on lib/travis_dedup.rb:
```diff
@@ -98,17 +100,36 @@ def host
       case response.status
       when 200 then JSON.parse(response.body)
       when 204 then nil
+      when 500
```
I will check Faraday code. Judging by the stackstrace (see https://github.com/grosser/travis_dedup/issues/9 ), I would say it does not raise on 500

grosser commented:
they told me the active build cancellation will come in Q1

On Fri, Nov 18, 2016 at 8:41 AM, Jean-Michel Garnier <
notifications@github.com> wrote:

> Thanks for the review.
> 
> Following your previous comment on the isue, I asked them about it. Their
> feature does a little bit less than travis_dedup so I'd rather move forward
> on this PR.
> 
> Main difference with travis_dedup is that it does not delete "Active"
> (running) jobs and we really value that behavior.
> 
> A screenshot about Travis feature:
> [image: Inline image 1]
> 
> On Fri, Nov 18, 2016 at 5:10 PM, Michael Grosser <notifications@github.com
> 
> > wrote:
> > 
> > FYI might not want to spend time here since travis is releasing dedup as
> > a
> > feature in Q1
> > 
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/grosser/travis_dedup/pull/11#issuecomment-261570374
> > ,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AAAfcyTpby6-
> > IcYNIMYWcfah3z1dtc01ks5q_c4NgaJpZM4K2bQV>
> > .
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/grosser/travis_dedup/pull/11#issuecomment-261578784,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAAsZ-681J4J78WBjs0RgswKeOk5jqfeks5q_dVMgaJpZM4K2bQV
> .

grosser commented on lib/travis_dedup.rb:
```diff
@@ -20,6 +23,8 @@ def cli(argv)
         BANNER
         opts.on("--pro", "travis pro") { self.pro = true }
         opts.on("--branches", "dedup builds on branches too") { self.branches = true }
+        opts.on("--retry", "number of times to retry when Travis returns a 500. #{RETRY} times by default") { |value| self.retry_option = value }
```
should be `--retry [COUNT], Integer, "number of ...` or it won't work

jmgarnier replied:
Good point. I did not test my code. Will fix it.

grosser commented on lib/travis_dedup.rb:
```diff
@@ -20,6 +23,8 @@ def cli(argv)
         opts.on("--pro", "travis pro") { self.pro = true }
         opts.on("--branches", "dedup builds on branches too") { self.branches = true }
+        opts.on("--retry", "number of times to retry when Travis returns a 500. #{RETRY} times by default") { |value| self.retry_option = value }
+        opts.on("--force", "give up cancelling duplicate builds when Travis keeps returning 500") { self.force = true }
```
really need that ?

jmgarnier replied:
When you have to restart manually 10 times a build because Travis API keeps returning 500 when running `travis_dedup`, we do need it, yes 😸

grosser commented:
that really sucks.... have you reached out to travis for this ?
... name it something like `ignore-500` or so ?

On Mon, Nov 21, 2016 at 9:09 AM, Jean-Michel Garnier <
notifications@github.com> wrote:

> ## _@jmgarnier_ commented on this pull request.
> 
> In lib/travis_dedup.rb https://github.com/grosser/travis_dedup/pull/11:
> 
> > @@ -20,6 +23,8 @@ def cli(argv)
> >          BANNER
> >          opts.on("--pro", "travis pro") { self.pro = true }
> >          opts.on("--branches", "dedup builds on branches too") { self.branches = true }
> > -        opts.on("--retry", "number of times to retry when Travis returns a 500. #{RETRY} times by default") { |value| self.retry_option = value }
> > -        opts.on("--force", "give up cancelling duplicate builds when Travis keeps returning 500") { self.force = true }
> 
> When you have to restart manually 10 times a build because Travis API
> keeps returning 500 when running travis_dedup, we do need it, yes 😸
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/grosser/travis_dedup/pull/11, or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAAsZyXkIwt-Km3tB3SiazScnABACN_Uks5rAdBPgaJpZM4K2bQV
> .

jmgarnier commented:
> that really sucks.... have you reached out to travis for this ?

Do you mean the fact that Travis API sometimes returns too many `500` in a row?
I reached their support when you first replied to my issue to ask about the release date.  But I did not mention it explicitly the fact the API is not stable. Will do tomorrow.

> name it something like `ignore-500` or so ?

Also, I have amended my commits, fixed the --retry option and renamed the option `--ignore-error-500`

grosser commented:
looks good!

grosser commented:
0.6.0 is out!

jmgarnier commented:
Wow, you were a bit too fast 😛 

I found a problem in a spec and amended my last commit at the same time you merged the PR.

 Will investigate later. Also I really wanted to test my forked gem to make sure code is really working as I may have forgotten something in the unit tests.

grosser commented:
but the code worked ? ... tests were green ...

anything big / should I revert or good to leave ?

On Mon, Nov 21, 2016 at 9:51 AM, Jean-Michel Garnier <
notifications@github.com> wrote:

> Wow, you were a bit too fast 😛
> 
> I found a problem in a spec and amended my last commit at the same time
> you merged the PR.
> 
> I really have to come home or my wife will become bersek;)
> 
> Will investigate later. Also I really wanted to test my forked gem to make
> sure code is really working as I may have forgotten something in the unit
> tests.
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> https://github.com/grosser/travis_dedup/pull/11#issuecomment-262014171,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAAsZ3xr3pO3gQiJU7yCZ_dGeKIdL99aks5rAdongaJpZM4K2bQV
> .

jmgarnier commented:
A quick FYI, Travis support got back to me and finally confirmed that the "auto-cancel running builds" feature will be shipped in Q1 :)
